docs/content/sips/000-app-config.md

```diff
@@ -0,0 +1,137 @@
+title = "SIP xxx - Application Configuration"
+template = "main"
+date = "2022-03-22T14:53:30Z"
+
+---
+
+Summary: A configuration system for Spin applications.
+
+Owner: lann.martin@fermyon.com
+
+Created: March 22, 2022
+
+Updated: April 1, 2022
+
+## Background
+
+It is common for applications to require configuration at runtime that isn't known at build time or is too sensitive to be stored with build artifacts:
+
+- Logging configuration
+- Per-channel (production, staging, etc) service dependency URLs
+- Database secrets
+
+## Proposal
+
+### Configuration is defined by components and applications
+
+Configuration within a "parent" (component or application) consists of a number of configuration "slots":
+- slots are uniquely identified within their parent by a string "key"; in order to allow unambiguous conversion to environment variables or file paths, keys are constrained:
+  - must start with a letter (required for env vars)
+  - consisting of only lowercase ascii alphanum and `_` (`[a-z0-9_]`)
+    - only one `_` at a time and not at the end (to allow delimiting in env vars with `__`)
+- a slot must _either_ be marked as "required" _or_ must be given a default value
+- a slot may be marked as "secret", in which case any associated value should be handled with care (e.g. not logged)
+
+```toml
+[config]
+# This simple form...
+key1 = "default_value"
+# ...is equivalent to:
+key1 = { default = "default_value" }
+# required & secret slot 
+key2 = { required = true, secret = true }
+```
+
+Defaults can use template strings to reference other slots.
+```toml
+key1 = { required = true }
+key2 = "prefix-{{ .key1 }}-suffix"
+```
+
+### Components and applications can set configuration of their direct dependencies
+
+In dependency configuration, templates can reference the app config and "ancestor" dependant configs:
+
+`spin.toml`:
+```toml
+[config]
+app_root = "/app"
+...
+[[component.config]]
+# Note: no '.'s needed when referencing top-level app config
+work_root = "{{ app_root }}/work"  # -> "/app/work"
+[[component.dependencies.dep1.config]]
+dep_root = "{{ ..work_root }}/dep" # -> "/app/work/dep"
```

Comment:
Do we need an explicit mention around how to reference 'ancestor' configs?

I see `..work_root` in this example, so it appears that a dep directly under an ancestor (one level down) can access an ancestor's config via `..<key>`.  Does it follow that a dep two levels down would use `...<key>`?  Or perhaps "two levels down" (or more) is not a valid use case and not supported...

Reply:
We don't even support dependencies yet, so I don't know exactly how this will end up working, but the code I'm working on allows any depth (up to a hard-coded limit of 1000 at the moment :slightly_smiling_face:).

These relative refs are a convenience to reduce duplication; ultimately all references will have to resolve to literal strings and top-level keys (i.e. those in `[config]`) and I expect that to be the most common way to use references.

Reply:
Ah but I see I don't mention it at all in the prose up here. Yeah I'll add something.

Reply:
OK, there are more words now.